### PR TITLE
fix(web): derive URLs from request headers for reverse-proxy support

### PR DESCRIPTION
**Pull request by xingyaoww**
## Problem

`openhands web` shows only the textual-serve splash screen (Textual logo + "uv run openhands") and never transitions to the actual TUI when accessed through a reverse proxy.

**Before (through proxy):**
image

The root cause is that textual-serve's `Server` class hardcodes all URLs using its `public_url`, which defaults to `http://0.0.0.0:PORT`. When served behind a reverse proxy:

- **Static assets** (`http://0.0.0.0:12000/static/js/textual.js`) → unreachable, JS never loads
- **WebSocket** (`ws://0.0.0.0:12000/ws`) → connection never establishes
- The intro screen stays visible forever because the JS that transitions to the terminal never loads

## Solution

Introduce `ProxyAwareServer`, a subclass of `textual_serve.server.Server`, that overrides `handle_index` to derive the base URL from the incoming HTTP request headers:

1. Reads `X-Forwarded-Proto` / `X-Forwarded-Host` headers (standard reverse-proxy headers)
2. Falls back to `request.scheme` / `request.host` for direct access
3. Constructs all URLs (static assets, WebSocket) from the derived base URL

**URL generation comparison:**

| Access method | Before | After |
|---|---|---|
| Via proxy (`https://proxy.example.com`) | `ws://0.0.0.0:12000/ws` ❌ | `wss://proxy.example.com/ws` ✅ |
| Direct (`http://localhost:12000`) | `ws://0.0.0.0:12000/ws` ❌ | `ws://localhost:12000/ws` ✅ |

## Changes

- `openhands_cli/tui/serve.py` — Added `ProxyAwareServer` class with request-aware URL generation; `launch_web_server` now uses it
- `tests/test_web_command.py` — Updated mocks to patch `ProxyAwareServer`; added 4 parametrized tests for `_get_base_url` covering direct access, full proxy headers, and partial header scenarios

## Verification

- ✅ Reproduced the issue (stuck splash screen via proxy URL)
- ✅ Verified fix works (full TUI loads with Settings dialog visible through proxy)
- ✅ `make lint` passes
- ✅ `make test` — all 1264 tests pass

---

## 🚀 Try this PR

```bash
uvx --python 3.12 git+https://github.com/OpenHands/OpenHands-CLI.git@fix/web-proxy-aware-urls
```

### CODE REVIEW

**Comment by xingyaoww** [comment]
@OpenHands https://github.com/OpenHands/OpenHands-CLI/pull/570#pullrequestreview-3895699951

address all reviews, respond to review thread with the commit ID when you've done addressing them, and close/resolve that particular thread

File: openhands_cli/tui/serve.py [hunk 1]
```diff
@@ -1,6 +1,63 @@
...
+            "app_websocket_url": get_websocket_url("websocket"),
+        }
+        context["config"] = {
+            "static": {
+                "url": get_url("static", filename="/").rstrip("/") + "/",
+            },
+        }
+        context["application"] = {
+            "name": self.title,
+        }
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands_cli/tui/serve.py]: Added a NOTE in the class docstring documenting this as a necessary full override of `Server.handle_index` (textual-serve 1.1.3) because URL generation uses local closures over `self.public_url` — there is no smaller hook to override. The note instructs maintainers to verify sync on textual-serve upgrades. Fixed in d5d1dbf.

File: openhands_cli/tui/serve.py [hunk 2]
```diff
@@ -1,6 +1,63 @@
...
+        font_size = to_int(request.query.get("fontsize", "16"), 16)
+
+        def get_url(route: str, **kwargs: str) -> str:
+            path = router[route].url_for(**kwargs)
+            return f"{base_url}{path}"
+
+        def get_websocket_url(route: str, **kwargs: str) -> str:
+            url = get_url(route, **kwargs)
+            if base_url.startswith("https"):
+                return "wss:" + url.split(":", 1)[1]
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands_cli/tui/serve.py]: Replaced with `urlparse`/`urlunparse` as suggested. Fixed in d5d1dbf.

File: openhands_cli/tui/serve.py [hunk 3]
```diff
@@ -1,6 +1,63 @@
...
+
+    @staticmethod
+    def _get_base_url(request: web.Request) -> str:
+        """Derive the external base URL from request headers.
+
+        Checks ``X-Forwarded-Proto`` / ``X-Forwarded-Host`` first (common
+        reverse-proxy headers), then falls back to the ``Host`` header.
+        """
+        proto = request.headers.get("X-Forwarded-Proto", request.scheme)
+        host = request.headers.get("X-Forwarded-Host", request.host)
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply openhands_cli/tui/serve.py:41]: Added validation: proto must be in {http, https} (falls back to `request.scheme`), host rejects empty, multi-line, or >253-char values (falls back to `request.host`). Also added 5 parametrized test cases covering these edge cases. Fixed in d5d1dbf.

File: tests/test_web_command.py [hunk 4]
```diff
@@ -113,3 +115,46 @@ def test_launch_web_server_propagates_exception(mock_server_class):
...
+        ),
+    ],
+)
+def test_get_base_url(headers, scheme, host, expected_base):
+    request = MagicMock()
+    request.headers = CIMultiDict(headers)
+    request.scheme = scheme
+    request.host = host
+
+    assert ProxyAwareServer._get_base_url(request) == expected_base
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply tests/test_web_command.py]: Added `handle_index` integration tests that wire a real aiohttp router and call the unwrapped handler (via `__wrapped__`) to verify the full context dict: WebSocket URLs, static asset URLs, application name, and fontsize query parameter — for both direct access and proxied scenarios. Fixed in d5d1dbf.

**Comment by xingyaoww** [comment]
@OpenHands pls fix https://github.com/OpenHands/OpenHands-CLI/actions/runs/22715405417/job/65863721069?pr=570

File: openhands_cli/tui/serve.py [hunk 5]
```diff
@@ -1,6 +1,84 @@
...
+            "app_websocket_url": get_websocket_url("websocket"),
+        }
+        context["config"] = {
+            "static": {
+                "url": get_url("static", filename="/").rstrip("/") + "/",
+            },
+        }
+        context["application"] = {
+            "name": self.title,
+        }
```
Comment (VascoSch92) [inline openhands_cli/tui/serve.py]: nit: This can be written just as one dictionary and returned directly

Comment (xingyaoww) [reply openhands_cli/tui/serve.py]: @OpenHands make it so!

**Comment by xingyaoww** [comment]
HUMAN: I've personally verified this works on my setup, so gonna merge this now
